File: interpreter/python/myinterpreter.py
class Item:
    def interpret(self, foodorder):
        pass

class FoodItem(Item):
    def __init__(self, name, type):
        self.name = name
        self.type = type

    def get_name(self):
        return self.name

    def get_type(self):
        return self.type
    
    def interpret(self, foodorder):
        if foodorder.get_type().lower() == "food":
            return f"Food Item: {foodorder.get_name()}"
        return None

class DrinkItem(Item):
    def __init__(self, name, size):
        self.name = name
        self.size = size

    def get_name(self):
        return self.name

    def get_size(self):
        return self.size

    def interpret(self, foodorder):
        if foodorder.get_type().lower() == "drink":
            return f"Drink Item: {foodorder.get_name()}"
        return None
    
class FoodOrder:
    def __init__(self, name, type):
        self.name = name
        self.type = type

    def get_name(self):
        return self.name

    def get_type(self):
        return self.type

class Interpreter:
    def __init__(self):
        self.items = []

    def add_item(self, item):
        self.items.append(item)

    def interpret(self, foodorder):
        for item in self.items:
            result = item.interpret(foodorder)
            if result is not None:
                return result
        return "Item not found in the order."

if __name__ == "__main__":  
    interpreter = Interpreter()
    interpreter.add_item(FoodItem("Burger", "food"))
    interpreter.add_item(DrinkItem("Coke", "large"))

    order1 = FoodOrder("Burger", "food")
    order2 = FoodOrder("Coke", "drink")
    order3 = FoodOrder("Fries", "food")

    print(interpreter.interpret(order1))  # Output: Food Item: Burger
    print(interpreter.interpret(order2))  # Output: Drink Item: Coke
    print(interpreter.interpret(order3))  # Output: Item not found in the order.

File: interpreter/swift/myinterpreter.swift
import Foundation

// Context: Holds the state of the order
class FoodOrder {
    var items: [String: String] = [:] // Map of FoodItem Name to FoodType
    
    func add(item: String, type: String) {
        items[item] = type
    }
}

// Abstract Expression
protocol Item {
    func interpret(order: FoodOrder) -> String
}

// Terminal Expression: Food
class FoodItem: Item {
    private var name: String
    
    init(name: String) {
        self.name = name
    }
    
    func interpret(order: FoodOrder) -> String {
        if let type = order.items[name] {
            return "Food: \(name) (\(type))"
        }
        return ""
    }
}

// Terminal Expression: Drink
class DrinkItem: Item {
    private var name: String
    
    init(name: String) {
        self.name = name
    }
    
    func interpret(order: FoodOrder) -> String {
        if let type = order.items[name] {
            return "Drink: \(name) (\(type))"
        }
        return ""
    }
}

// Non-terminal Expression: Combines Food and Drink
class AllFood: Item {
    private var food: FoodItem
    private var drink: DrinkItem
    
    init(food: FoodItem, drink: DrinkItem) {
        self.food = food
        self.drink = drink
    }
    
    func interpret(order: FoodOrder) -> String {
        return "\(food.interpret(order: order)) and \(drink.interpret(order: order))"
    }
}

// --- Main Execution ---

let myOrder = FoodOrder()
myOrder.add(item: "Burger", type: "Main Course")
myOrder.add(item: "Cola", type: "Beverage")

let burger = FoodItem(name: "Burger")
let soda = DrinkItem(name: "Cola")

// Interpret individual items
print(burger.interpret(order: myOrder)) // Output: Food: Burger (Main Course)

// Interpret combined expression
let combo = AllFood(food: burger, drink: soda)
print("Combo Order: \(combo.interpret(order: myOrder))") 
// Output: Combo Order: Food: Burger (Main Course) and Drink: Cola (Beverage)
